File: heartbleed.c
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>

uint8_t *parse(
  uint8_t *ptr_request,
  uint32_t request_size,
  uint32_t *ptr_payload_length,
  uint8_t *ptr_error_code
) {
    uint8_t *response = (uint8_t *)malloc(sizeof(uint8_t) * (request_size - 2));
    if (response == NULL) {
        // puts("response-malloc is failed.");
        *ptr_error_code = 1;
        return response;
    }
    *ptr_payload_length = ((uint32_t)ptr_request[0] << 8) | (uint32_t)ptr_request[1];
    memcpy(response, ptr_request + 2, *ptr_payload_length);
    return response;
}

int main(int argc, char const *argv[])
{
    // 0x0f, 0xa0 = 4000 (payload_length)
    // 0x70, 0x61, 0x79, 0x6c, 0x6f, 0x61, 0x64 = "payload" (payload)
    uint8_t request[] = {0x0f, 0xa0, 0x70, 0x61, 0x79, 0x6c, 0x6f, 0x61, 0x64};
    // uint8_t request[] = {0x00, 0x07, 0x70, 0x61, 0x79, 0x6c, 0x6f, 0x61, 0x64};
    size_t request_size = sizeof(request) / sizeof (uint8_t);
    if (request_size < 3 || request_size > 65535 + 2) {
        printf("request_size is invalid.\nrequest_size = %zu\n",request_size);
        return 1;
    }

    uint8_t error_code = 0;
    uint32_t payload_length = 0;
    uint8_t *response = parse(request, request_size, &payload_length, &error_code);
    if (error_code != 0) {
        if (error_code == 1)
            puts("response-malloc is failed.");
        else if (error_code == 2)
            puts("packet-parse is failed.");
        else
            puts("unexpected error.");
        return 1;
    }

    for (uint32_t i = 0; i < payload_length; i++)
        printf("%c", response[i]);
    free(response);
    return 0;
}

File: heartbleed.rs

// fn parse(
//     ptr_request: &[u8],
//     request_size: usize,
//     mut ptr_payload_length: u32,
//     ptr_error_code: &u8
//  ) {
//     //  let a = request_size - 2;
//      let mut response: Vec<u8> = Vec::with_capacity(request_size - 2);
//     // let temp: u32 = ((ptr_request[0] as u32) << 8 | (ptr_request[1] as u32));
//     ptr_payload_length = ((ptr_request[0] as u32) << 8 | (ptr_request[1] as u32));
//     response.extend(ptr_request)
// }

fn main() {
    let request = vec![0x00, 0x07, 0x70, 0x61, 0x79, 0x6c, 0x6f, 0x61, 0x64];
    let request_size: usize = request.len();
    if request_size < 3 ||  request_size > 65535 + 2 {
        println!("request_size is invalid.\nrequest_size = {}\n", request_size);
        std::process::exit(1);
    }

    let mut response: Vec<u8> = Vec::with_capacity(request_size - 2);
    let payload_length: u32 = (request[0] as u32) << 8 | (request[1] as u32);
    response.extend(request[2..2+payload_length as usize].iter().cloned());

    for i in 0..payload_length {
        print!("{}", response[i as usize] as char);
    }
    std::process::exit(0);
 }